lucene/core/src/java/org/apache/lucene/search/MaxScoreTerminator.java

```diff
@@ -0,0 +1,170 @@
+ *
+ * <p>Once the total number of documents collected reaches the requested total (numHits),
+ * the worst-scoring leaf can no longer contribute any documents to the results, so it can be
+ * terminated, and any leaves whose scores rise above that worst score is no longer competitive and
+ * can also be terminated. If we kept a global priority queue we could update the global maximum
+ * competitive score, and use that as a termination threshold, but assuming this to be too costly
+ * due to thread contention, we seek to more cheaply update an upper bound on the worst score.
+ * Specifically, when a leaf is terminated, if the remaining leaves together have collected >= numHits,
+ * then we can update the maximum to the max of *their* max scores, excluding the terminated leaf's max
+ * from consideration.</p>
+ *
+ * <p> In practice this leads to a good bound on the number of documents collected, which tend to
+ * exceed numHits by a small factor.  When the documents are evenly distributed among N segments,
+ * we expect to collect approximately (N+1/N) * numHits documents. In a worst case, where *all*
+ * the best documents are in a single segment, we expect to collect something O(log N) ie
+ * (1/N + 1/N-1 + ... + 1) * numHits documents, which is still much better than
+ * the N * numHits we would collect with a naive strategy.
+ * </p>
+ */
+class MaxScoreTerminator {
+
+  // we use 2^5-1 to check the remainder with a bitwise operation
+  // private static final int DEFAULT_INTERVAL_BITS = 10;
+  private static final int DEFAULT_INTERVAL_BITS = 2;
+  private int interval;
+  int intervalMask;
+
+  /** The worst score for each leaf */
+  private final List<LeafState> leafStates;
+
+  /** The total number of docs to collect: from the Collector's numHits */
+  final int totalToCollect;
+
+  /** An upper bound on the number of docs "excluded" from max-score accounting due to early termination. */
+  private int numExcludedBound;
+
+  /** A lower bound on the total hits collected by all leaves */
+  int totalCollected;
+
+  /** the worst hit over all */
+  LeafState leafState;
+
+  MaxScoreTerminator(int totalToCollect) {
+    leafStates = new ArrayList<>();
+    this.totalToCollect = totalToCollect;
+    setIntervalBits(DEFAULT_INTERVAL_BITS);
+  }
+
+  synchronized LeafState add() {
+    LeafState newLeafState = new LeafState();
+    leafStates.add(newLeafState);
+    if (leafState == null) {
+      leafState = newLeafState;
+    }
+    return newLeafState;
+  }
+
+  // for testing
+  void setIntervalBits(int bitCount) {
+    interval = 1 << bitCount;
+    intervalMask = interval - 1;
+  }
+
+  /**
+   * Called by leaf collectors periodically to update their progress.
+   * @param newLeafState the leaf collector's current lowest score
+   * @return whether the collector should terminate
+   */
+  synchronized boolean update(LeafState newLeafState) {
+    totalCollected += interval;
```

Comment:
Yes it's true that calling threads must abide by the calling frequency, but the distinction between competitive and noncompetitive hits is moot since it will never be called in the non-competitive case: if the leaf knows its hit is noncompetitive, it is expected to terminate. I'll beef up the javadocs here